### PR TITLE
Partial migration from Travis to GitHub Actions based on Docker images

### PR DESCRIPTION
**Pull request by GreyCat**
This change partially migrates Kaitai Struct per-target test runs from using Travis to GitHub Actions based on [custom-built Docker images](https://github.com/orgs/kaitai-io/packages) originating from [kaitai_struct_docker_images repo](https://github.com/kaitai-io/kaitai_struct_docker_images).

This partially addresses https://github.com/kaitai-io/kaitai_struct/issues/824

## Pros

The primary advantage of this approach is being able to reproduce the exact same build locally, outside of any CI. This makes any kind of custom "environment preparation" scripts/methods (such as custom YAML in .travis.yml, or custom actions in GitHub Actions workflow YAML, or complicated and fragile stuff like `/prepare-$TARGET` scripts) obsolete and unnecessary.

## Cons

Potential downside is the speed of such CI. Thankfully, looks like at least for GitHub Actions implementation, this is not big concern — dockerized builds (which literally just download and run the container) vs traditional builds (which take image provided by CI owners + tune it by installing/enabling certain things on top of that image on the fly) work with very comparable speed.

## Speed comparison

I've compiled some speed comparison table for reference.

Data sources:

* Travis build: https://app.travis-ci.com/github/kaitai-io/ci_targets/builds/264105892
* Previous traditional build: https://github.com/kaitai-io/ci_targets/actions/runs/5406194401/usage
* Latest dockerized build: https://github.com/kaitai-io/ci_targets/actions/runs/5461386269/usage

Target/subtarget | Variety | Travis | GH Trad | GH Docker
-- | -- | -- | -- | --
cpp_stl_98 | gcc4.8_linux | 192 |   | 167
cpp_stl_98 | clang3.5_linux | 193 |   |  
cpp_stl_98 | clang7.3_osx | 255 |   |  
cpp_stl_11 | gcc4.8_linux | 313 |   | 205
cpp_stl_11 | clang3.5_linux | 396 |   |  
cpp_stl_11 | clang7.3_osx | 335 |   |  
java | openjdk7 | 49 | 51 | 45
java | oraclejdk8 | 58 | 53 | 65
java | openjdk11 | 51 | 51 | 54
java | openjdk17 | 97 | 48 | 61
javascript | nodejs4 | 45 |   | 18
javascript | nodejs8 | 51 |   | 20
javascript | nodejs10 | 49 |   |  
javascript | nodejs12 | 64 |   | 18
javascript | nodejs16 | 49 |   | 15
perl | 5.24 | 128 |   |  
perl | 5.36 | 112 |   |  
php | 7.1 | 69 |   |  
php | 8 | 58 |   |  
python | 2.7 | 54 |   | 38
python | 3.4 | 54 |   | 37
python | 3.1 | 62 |   |  
construct | python2.7 | 56 |   |  
construct | python3.4 | 49 |   |  
construct | python3.10 | 55 |   |  
ruby | 1.9 | 119 |   | 36
ruby | 2.3 | 79 |   |  
ruby | 3 | 38 |   | 21
go | 1.13 | 62 |   |  
go | 1.18 | 62 |   |  
rust | 1.31.1 | 317 |   |  
lua | 5.3 | 55 |   |  
csharp | mono5.18.0 | 118 |   |  
graphviz | 2.36 | 52 |   |  
nim | 1.6.0 | 235 |   |  

TL/DR: Travis->GH improves build times (sometimes slightly, sometimes dramatically), GH Trad->GH Docker has some slight negative effect, which, in my opinion, is totally ok given the flexibility we get.

## Additional changes

There are a few additional changes worth mentioning associated with this PR as well:

* Travis' cpp_stl/gcc4.8_linux, despite what the name suggest, haven't been "gcc4.8" for a very long time. This image was  upgraded to have gcc 5.4.0, so in reality that's what we were testing last few years. We're switching to true gcc4.8.
* cpp_stl_98 on gcc11 is added.
* cpp_stl_11 on gcc11 could be added too, but currently results in an error that I will address separately, so commented out for now.
* Java JDK build suppliers change with this, but, realistically, that's where the industry has been heading for last few years. We can't avoid this anymore.
* Ruby 3.0 -> 3.2, as that's the current modern version.
* Python 3.10 -> 3.11, as that's the current modern version.

## Future work to address

* Travis' clang3.5_linux is actually supplying clang version 7.0.0.
* Travis' clang7.3_osx is actually supplying "Apple LLVM version 9.1.0 (clang-902.0.39.2), Target: x86_64-apple-darwin17.7.0"
* https://ci.kaitai.io/ has to be updated for all version bumps

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -97,3 +72,128 @@ jobs:
...
+            subtarget: _11
+            implementation: gcc4.8
+          - target: cpp_stl
+            subtarget: _98
+            implementation: gcc11
+#          - target: cpp_stl
+#            subtarget: _11
+#            implementation: gcc11
+          - target: java
+            subtarget: ''
```
Comment (generalmimon) [inline .github/workflows/main.yml]: It seems that it's actually possible to omit these empty string settings, which would shorten the config a bit if we don't need any meaningful subtarget for these jobs. If you see some value in keeping them, feel free to - I'm just noting they're not necessary.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -97,3 +72,128 @@ jobs:
...
+          - target: ruby
+            subtarget: ''
+            implementation: 1.9
+          - target: ruby
+            subtarget: ''
+            implementation: 2.3
+          - target: ruby
+            subtarget: ''
+            implementation: 3.2
+    steps:
```
Comment (generalmimon) [inline .github/workflows/main.yml:125]:
```suggestion
    env:
      TARGET: ${{matrix.tsv.target}}
      SUBTARGET: ${{matrix.tsv.subtarget}}
      IMPL: ${{matrix.tsv.implementation}}
    steps:
```

```diff
       - name: docker pull
         run: |
-          docker pull ghcr.io/kaitai-io/kaitai-${{matrix.tsv.target}}-${{matrix.tsv.implementation}}-linux-x86_64
+          docker pull ghcr.io/kaitai-io/kaitai-"$TARGET"-"$IMPL"-linux-x86_64
```

```diff
       - name: download runtime
         run: |
           mkdir -p runtime
-          git clone https://github.com/kaitai-io/kaitai_struct_${{matrix.tsv.target}}_runtime runtime/${{matrix.tsv.target}}
+          git clone https://github.com/kaitai-io/kaitai_struct_"$TARGET"_runtime runtime/"$TARGET"
```

```diff
         run: |
           cd tests
-          ./docker-ci -t "${{matrix.tsv.target}}" -u "${{matrix.tsv.subtarget}}" -i "${{matrix.tsv.implementation}}"
+          ./docker-ci -t "$TARGET" -u "$SUBTARGET" -i "$IMPL"
       # - name: publish as artifact
       #   uses: actions/upload-artifact@v3
       #   with:
       #     name: test-${{matrix.tsv.target}}${{matrix.tsv.subtarget}}-${{matrix.tsv.implementation}}
       #     path: tests/test_out
       - name: publish
         env:
           BOT_SSH_KEY: ${{secrets.BOT_SSH_KEY}}
-          TARGET: ${{matrix.tsv.target}}
-          SUBTARGET: ${{matrix.tsv.subtarget}}
-          VARIETY: ${{matrix.tsv.implementation}}
         run: |
           ./push_artifacts/git_config_kaitai_bot
           ./push_artifacts/publish \
             -o kaitai-io \
             -r ci_artifacts \
             -m "Build results of ${GITHUB_REF#refs/heads/*} kaitai-io/ci_targets@$GITHUB_SHA" \
-            -b "$TARGET$SUBTARGET/$VARIETY" \
+             -b "${TARGET}${SUBTARGET}"/"$IMPL" \
             -- --exclude=.git --exclude=.travis.yml tests/test_out
```

Comment (generalmimon) [reply .github/workflows/main.yml:125]: @GreyCat Also check out https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections

File: .github/workflows/main.yml (hunk 1 above)
Comment (GreyCat) [reply .github/workflows/main.yml]: Very good point, thanks! Let me try that.

File: .github/workflows/main.yml (hunk 2 above)
Comment (GreyCat) [reply .github/workflows/main.yml:125]: Thanks! On the security aspects of it, I believe I've set up this repo to have any PR requests build to be approved by Kaitai team members. So, if we see that somebody wants to do something fishy with anything in the pipeline in relation to BOT_SSH_KEY, we really don't want to approve launching it.

**Pull request merged**